cupy/cuda/cupy_thrust.cu

```diff
@@ -225,51 +233,59 @@ bool _real_less(const T& lhs, const T& rhs) {
 
 // specialize thrust::less for float
 template <>
-__host__ __device__ __forceinline__
-#if (__CUDACC_VER_MAJOR__ >11 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2) || HIP_VERSION >= 402)
-THRUST_OPTIONAL_CPP11_CONSTEXPR_LESS
-#endif
-bool thrust::less<float>::operator() (
-    const float& lhs, const float& rhs) const {
+struct thrust::less<float> {
```

Comment:
important: I would think that we'd consider adding user-defined specializations for `thrust::less<float>`  to still be invalid.

For example, I believe in `thrust::sort` we dispatch to a merge sort vs radix sort implementation based on the operator. If the operator is a well known operator for a well known type (like thrust::less<float>).

If a user provides their own specialization of that operator, then that dispatch would no longer be valid because the radix sort code path wouldn't actually use their operator specialization.

TL;DR: I think the easiest solution here is to change this code from using and specializing `thrust::less` directly and just add a custom comparison type (like `cupy::less`) and pass that into the `thrust::sort` call.

Reply:
> TL;DR: I think the easiest solution here is to change this code from using and specializing `thrust::less` directly and just add a custom comparison type (like `cupy::less`) and pass that into the `thrust::sort` call.

Yes this sounds good to me.

Reply:
As I suspected, calling `thrust::sort(..., thrust::less<float>())` with a custom specialization of `thrust::less<float>` *will never invoke that specialization*. It dispatches to the RadixSort code path.

So cupy is just getting lucky that CUB's RadixSort happens to order NaNs in a way that matches whatever their tests expect.

https://godbolt.org/z/scezav1Yv

Same story for `thrust::stable_sort`: https://godbolt.org/z/n56EPbo3Y

TL;DR: This is all the more reason that specializing `thrust::less` is incorrect and cupy needs to define it's own custom `cupy::less` operator.

Reply:
> So cupy is just getting lucky that CUB's RadixSort happens to order NaNs in a way that matches whatever their tests expect.

This is actually not true. Let's discuss this example offline tomorrow when Bernhard is up:
```python
>>> a = cp.random.random((3, 5), dtype=cp.float32)
>>> cp.sort(a, axis=-1)
```

Reply:
> important: I would think that we'd consider adding user-defined specializations for `thrust::less<float>` to still be invalid.

Yes we do. But continuing to provide a specialization for `thrust::less` seemed the least different change and it would at least work with the planned changes in CCCL.

> For example, I believe in `thrust::sort` we dispatch to a merge sort vs radix sort implementation based on the operator. If the operator is a well known operator for a well known type (like thrust::less).
> 
> If a user provides their own specialization of that operator, then that dispatch would no longer be valid because the radix sort code path wouldn't actually use their operator specialization.

This is what's probably happening here. 

> TL;DR: I think the easiest solution here is to change this code from using and specializing `thrust::less` directly and just add a custom comparison type (like `cupy::less`) and pass that into the `thrust::sort` call.

I had to do a more complicated version, because we only want a custom less function object for the cases where the comparison semantics are changed. For integers, we would still want to use `thrust::less` to benefit, e.g., from radix sort.